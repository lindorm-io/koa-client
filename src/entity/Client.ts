import { ClientEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions, TObject } from "@lindorm-io/core";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  extra: TObject<any>;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  extra?: TObject<any>;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _extra: TObject<any>;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._extra = options?.extra || {};
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get extra(): TObject<any> {
    return this._extra;
  }
  public set extra(object: TObject<any>) {
    this._extra = object;
    this.addEvent(ClientEvent.EXTRA_CHANGED, { extra: this._extra });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new EntityCreationError("Client");
    }
    this.addEvent(ClientEvent.CREATED, {
      approved: this._approved,
      description: this._description,
      extra: this._extra,
      name: this._name,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
