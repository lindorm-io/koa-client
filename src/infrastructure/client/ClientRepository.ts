import { Client, IClient } from "../../entity";
import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { MongoCollection } from "../../enum";
import { indices } from "./indices";
import { schema } from "./schema";

export interface IClientFilter {
  id?: string;
  approved?: boolean;
  name?: string;
}

export interface IClientRepository extends IRepository<Client> {
  create(entity: Client): Promise<Client>;
  update(entity: Client): Promise<Client>;
  find(filter: IClientFilter): Promise<Client>;
  findMany(filter: IClientFilter): Promise<Array<Client>>;
  findOrCreate(filter: IClientFilter): Promise<Client>;
  remove(entity: Client): Promise<void>;
}

export class ClientRepository extends RepositoryBase<Client> implements IClientRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.CLIENT,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: IClient): Client {
    return new Client(data);
  }

  protected getEntityJSON(entity: Client): IClient {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      approved: entity.approved,
      description: entity.description,
      extra: entity.extra,
      name: entity.name,
      secret: entity.secret,
    };
  }

  public async create(entity: Client): Promise<Client> {
    return super.create(entity);
  }

  public async update(entity: Client): Promise<Client> {
    return super.update(entity);
  }

  public async find(filter: IClientFilter): Promise<Client> {
    return super.find(filter);
  }

  public async findMany(filter: IClientFilter): Promise<Array<Client>> {
    return super.findMany(filter);
  }

  public async findOrCreate(filter: IClientFilter): Promise<Client> {
    return super.findOrCreate(filter);
  }

  public async remove(entity: Client): Promise<void> {
    await super.remove(entity);
  }
}
