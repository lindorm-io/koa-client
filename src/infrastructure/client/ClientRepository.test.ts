import MockDate from "mockdate";
import { Client } from "../../entity";
import { ClientRepository } from "./ClientRepository";
import { Logger, LogLevel } from "@lindorm-io/winston";
import { baseHash } from "@lindorm-io/core";
import {
  IMongoConnectionOptions,
  MongoConnection,
  MongoConnectionType,
  RepositoryEntityNotFoundError,
} from "@lindorm-io/mongo";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

const mongoOptions: IMongoConnectionOptions = {
  type: MongoConnectionType.MEMORY,
  auth: {
    user: "user",
    password: "password",
  },
  url: {
    host: "host",
    port: 1234,
  },
  databaseName: "databaseName",
};

const logger = new Logger({ packageName: "n", packageVersion: "v" });
logger.addConsole(LogLevel.ERROR);

describe("ClientRepository", () => {
  let repository: ClientRepository;
  let client: Client;

  beforeEach(async () => {
    const mongo = new MongoConnection(mongoOptions);
    await mongo.connect();

    repository = new ClientRepository({
      db: mongo.getDatabase(),
      logger,
    });
    client = new Client({
      approved: true,
      description: "description",
      extra: { emailAuthorizationUri: "https://lindorm.io" },
      name: "name",
      secret: baseHash("secret"),
    });
  });

  test("should create", async () => {
    await expect(repository.create(client)).resolves.toMatchSnapshot();
  });

  test("should update", async () => {
    await repository.create(client);

    await expect(repository.update(client)).resolves.toMatchSnapshot();
  });

  test("should find", async () => {
    await repository.create(client);

    await expect(repository.find({ name: "name" })).resolves.toMatchSnapshot();
  });

  test("should remove", async () => {
    await repository.create(client);

    await expect(repository.remove(client)).resolves.toBe(undefined);
    await expect(repository.find({ id: client.id })).rejects.toStrictEqual(expect.any(RepositoryEntityNotFoundError));
  });
});
