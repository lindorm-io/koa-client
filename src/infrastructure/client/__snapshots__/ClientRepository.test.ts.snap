// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientRepository should create 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "client_created",
      "payload": Object {
        "approved": true,
        "created": 2020-01-01T07:00:00.000Z,
        "description": "description",
        "extra": Object {
          "emailAuthorizationUri": "https://lindorm.io",
        },
        "name": "name",
        "secret": "c2VjcmV0",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  ],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`ClientRepository should find 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "client_created",
      "payload": Object {
        "approved": true,
        "created": 2020-01-01T07:00:00.000Z,
        "description": "description",
        "extra": Object {
          "emailAuthorizationUri": "https://lindorm.io",
        },
        "name": "name",
        "secret": "c2VjcmV0",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  ],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`ClientRepository should update 1`] = `
Client {
  "_approved": true,
  "_created": 2020-01-01T07:00:00.000Z,
  "_description": "description",
  "_events": Array [
    Object {
      "date": 2020-01-01T07:00:00.000Z,
      "name": "client_created",
      "payload": Object {
        "approved": true,
        "created": 2020-01-01T07:00:00.000Z,
        "description": "description",
        "extra": Object {
          "emailAuthorizationUri": "https://lindorm.io",
        },
        "name": "name",
        "secret": "c2VjcmV0",
        "updated": 2020-01-01T07:00:00.000Z,
      },
    },
  ],
  "_extra": Object {
    "emailAuthorizationUri": "https://lindorm.io",
  },
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 1,
}
`;
